lib/publiccloud/provider.pm

```diff
@@ -705,22 +705,36 @@ sub terraform_param_tags
     return encode_json($tags);
 }
 
-=head2 get_terraform_output
+=head2 get_terraform_output($jq_query, $tf_dir)
 
 Query the terraform data structure in json format.
+Automatic search and set valid terraform directory on helper image.
 Input: <jq-query-format> string; <empty> = no query then full output of data structure.
+       <tf_dir> string, optional terraform destination directory; when empty, dir.searched.
 E.g: to get the VM instance name from json data structure, the call is: 
     get_terraform_output(".vm_name.value[0]");
 To get the complete output structure, the call is:
-    get_terraform_output();
+    get_terraform_output(".");
 
 =cut
 
 sub get_terraform_output {
-    my ($self, $jq_query) = @_;
-    my $res = script_output("terraform output -no-color -json | jq -r '$jq_query' 2>/dev/null", proceed_on_failure => 1);
-    # jq 'null' shall return empty
-    return $res unless ($res =~ /^null$/);
+    my ($self, $jq_query, $tf_dir) = @_;
+    # cuurent dir saved
+    my $dir = script_output("pwd");
+    # terraform state file search first matching
+    my $start = ($tf_dir) ? $tf_dir : '~';
+    $tf_dir = script_output("find $start -name \*.tfstate -exec dirname {} + -quit 2>/dev/null");
+    # state file missing: no data found
+    return "" if ($tf_dir eq "");
+    # enter terraform dir. found
+    script_run("cd $tf_dir") if ($tf_dir ne $dir);
+    # get data
+    my $res = script_output("terraform output -no-color -json | jq -Mr '$jq_query'", proceed_on_failure => 1);
```

Comment:
Why send this via a pipe to `jq` when you could upload all the json output in a variable directly in Perl and do `decode_json`? Judging by L705 the JSON modules are already imported in this library.

Reply:
Yes, it was an idea I also considered.
But here it is intended as first-version simplification, because we needed basically a string as output value, the `instance-id name`, also allowing general usage to return other json structures, on the need.

Also, the routine should handle more  additional perl commands to achive the result from then total json.

But pls consider that in this routine version, the full json output could be achieved using "." as input query, that could be assigned to a variable.

So the scope was first 1) automatic search (`find`) of the right place where terraform data present, unless input-directory provided; then 2) validation of the place (is present a *.tfstate file); 3) run command in that place, to get data based on query; 4) return ouptut or empty. 

So eventually your note would be a discussion on the phase n.4, also welcome if needed.